075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+Their values are got from the new `KafkaCluster.addedNodes` method (needs to be implemented) and the already existing `KafkaCluster.removedNodes` one.
+
+### Auto-rebalancing execution
+
+Before describing the process in more details, it's worth mentioning how scaling up and down happens today within the Cluster Operator reconciliation flow.
+
+The scaling up is triggered asynchronously during a reconciliation which isn't blocked on waiting for the newly added nodes to be ready.
+In the `KafkaReconcile.reconcile()` method, it is triggered via the `podSet()` method which updates the involved `StrimziPodSet`(s) so that the corresponding controller will take care of spinning up the newly added nodes while the current reconciliation just ends.
+It means that, after the scaling up is triggered, the new pods will take time to be up, running and ready, so it's likely that the auto-rebalancing will happen on a future reconciliation.
+
+The scaling down runs synchronously within a single reconciliation which is blocked on waiting for the nodes to be removed.
+In the `KafkaReconcile.reconcile()` method, it is executed via the `scaleDown()` method which updates the involved `StrimziPodSet`(s) so that the corresponding controller will take care of removing the nodes while the current reconciliation waits for the end of the operation.
+
+The idea is about adding a new `autoRebalancing()` method in the `KafkaReconcile.reconcile()` flow, taking care of starting a rebalancing and checking the status of an ongoing one across reconciliations.
+The `KafkaReconcile.reconcile()` flow, taking into account methods involved in the scaling and rebalancing, would be something like:
+
+* ... 
+* `scaleDown()`
+* ...
+* `podSet()`
+* ...
+* `autoRebalancing()`
+* ...
+
+> In case of scaling down, the rebalancing has to happen before removing brokers and not after that as it looks like from the above flow. Despite that, some internal state will help to have the right actions order across reconciliations as explained in the next sections.
+
+#### Auto-rebalancing on scaling up
+
+The following flow is triggered when the user increases the `replicas` field within a `Kafka` or `KafkaNodePool` custom resource in order to scale up the cluster by adding more brokers.
+
+1. On the **1st** reconciliation (scaling up is requested): 
+     * the `podSet()` method triggers the scaling up, as described in the previous section, but the reconciliation is not blocked on waiting for the newly added pods to be ready.
+     * the `autoRebalancing()` method detects that a scaling up was requested because of `KafkaCluster.addedNodes()` returning a non empty list with the brokers' IDs to be added to the cluster. Because the `Kafka.status.cruiseControl.autoRebalance.state` doesn't exist, it's a new operation so the state is updated as `ScalingUp` and the `add-brokers` with the brokers' IDs. Of course, it doesn't trigger any rebalance because the scaling up is going on asynchronously.
```

Comment:
Lot of information on a single line, so it is hard to comment ...
* What if the `Kafka.status.cruiseControl.autoRebalance.state` exists?
* What does it do if it `doesn't trigger any rebalance because the scaling up is going on asynchronously.`?

Reply:
> What if the Kafka.status.cruiseControl.autoRebalance.state exists?

Yeah, as Fede pointed out, having a state there means that we are not in this first reconciliation. We are running one of the next or anyway recovering the actions if the operator fails before (i.e. it crashed).

> What does it do if it doesn't trigger any rebalance because the scaling up is going on asynchronously.?

It just updates the state to "ScalingUp" and update the "add-brokers" list (as described) to be ready when scaling up ends and it's time to start the rebalancing.